services/galley/src/Galley/Data.hs

```diff
@@ -85,17 +108,131 @@ import qualified System.Logger.Class  as Log
+        Nothing -> paginate Cql.selectUserTeams (paramsP Quorum (Identity usr) (max + 1))
+  where
+    strip p = p { result = take (fromIntegral max) (result p) }
+
+teamConversation :: MonadClient m => TeamId -> ConvId -> m (Maybe TeamConversation)
+teamConversation t c = fmap (newTeamConversation c . runIdentity) <$>
+    retry x1 (query1 Cql.selectTeamConv (params Quorum (t, c)))
+
+teamConversations :: MonadClient m => TeamId -> m [TeamConversation]
+teamConversations t = map (uncurry newTeamConversation) <$>
+    retry x1 (query Cql.selectTeamConvs (params Quorum (Identity t)))
+
+teamMembers :: MonadClient m => TeamId -> m [TeamMember]
+teamMembers t = map (uncurry newTeamMember) <$>
+    retry x1 (query Cql.selectTeamMembers (params Quorum (Identity t)))
+
+teamMember :: MonadClient m => TeamId -> UserId -> m (Maybe TeamMember)
+teamMember t u = fmap (newTeamMember u . runIdentity) <$>
+    retry x1 (query1 Cql.selectTeamMember (params Quorum (t, u)))
+
+userTeams :: MonadClient m => UserId -> m [TeamId]
+userTeams u = map runIdentity <$>
+    retry x1 (query Cql.selectUserTeams (params Quorum (Identity u)))
+
+createTeam :: MonadClient m
+           => UserId
+           -> Range 1 256 Text
+           -> Range 1 256 Text
+           -> Maybe (Range 1 256 Text)
+           -> m Team
+createTeam uid (fromRange -> n) (fromRange -> i) k = do
+    tid <- Id <$> liftIO nextRandom
+    retry x5 $ write Cql.insertTeam (params Quorum (tid, uid, n, i, fromRange <$> k))
+    pure (newTeam tid uid n i & teamIconKey .~ (fromRange <$> k))
+
+deleteTeam :: MonadClient m => TeamId -> m ()
+deleteTeam tid = do
+    retry x5 $ write Cql.markTeamDeleted (params Quorum (Identity tid))
+    mm <- teamMembers tid
+    for_ mm $ removeTeamMember tid . view userId
+    cc <- teamConversations tid
+    for_ cc $ removeTeamConv tid . view conversationId
+    retry x5 $ write Cql.deleteTeam (params Quorum (Identity tid))
+
+addTeamMember :: MonadClient m => TeamId -> TeamMember -> m ()
+addTeamMember t m =
+    retry x5 $ batch $ do
+        setType BatchLogged
+        setConsistency Quorum
+        addPrepQuery Cql.insertTeamMember (t, m^.userId, m^.permissions)
+        addPrepQuery Cql.insertUserTeam   (m^.userId, t)
+
+removeTeamMember :: MonadClient m => TeamId -> UserId -> m ()
+removeTeamMember t m =
+    retry x5 $ batch $ do
+        setType BatchLogged
+        setConsistency Quorum
+        addPrepQuery Cql.deleteTeamMember (t, m)
+        addPrepQuery Cql.deleteUserTeam   (m, t)
+
+removeTeamConv :: MonadClient m => TeamId -> ConvId -> m ()
+removeTeamConv tid cid = do
+    deleteConversation cid
+    retry x5 $ batch $ do
+        setType BatchLogged
+        setConsistency Quorum
+        addPrepQuery Cql.deleteTeamConv (tid, cid)
+
+updateTeam :: MonadClient m => TeamId -> TeamUpdateData -> m ()
+updateTeam tid u = retry x5 $ batch $ do
+    setType BatchLogged
```

Comment:
I chose to use it because of the coupling between `icon` and `icon_key` updates.